src/stream/src/executor/temporal_filter.rs

```diff
@@ -0,0 +1,466 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::BTreeMap;
+use std::ops::Deref;
+
+use chrono::NaiveDateTime;
+use futures::StreamExt;
+use futures_async_stream::try_stream;
+use itertools::Itertools;
+use risingwave_common::array::{ArrayImpl, DataChunk, Op, Row, StreamChunk};
+use risingwave_common::buffer::{Bitmap, BitmapBuilder};
+use risingwave_common::catalog::Schema;
+use risingwave_common::must_match;
+use risingwave_common::types::{DataType, IntervalUnit, NaiveDateTimeWrapper, ScalarImpl};
+use risingwave_common::util::chunk_coalesce::DataChunkBuilder;
+use risingwave_common::util::epoch::Epoch;
+use risingwave_common::util::select_all;
+use risingwave_expr::expr::expr_binary_nonnull::new_binary_expr;
+use risingwave_expr::expr::{BoxedExpression, InputRefExpression, LiteralExpression};
+use risingwave_expr::vector_op::arithmetic_op::timestamp_interval_sub;
+use risingwave_expr::ExprError;
+use risingwave_pb::expr::expr_node::Type;
+use risingwave_storage::table::streaming_table::state_table::StateTable;
+use risingwave_storage::StateStore;
+
+use super::utils::get_rowstreams_by_new_vnodes;
+use super::{
+    expect_first_barrier, ActorContextRef, BoxedExecutor, BoxedMessageStream, Executor, Message,
+    PkIndices, PkIndicesRef, StreamExecutorError, StreamExecutorResult,
+};
+use crate::common::InfallibleExpression;
+
+/// [`TemporalBufferKey`] contains a record's timestamp and pk.
+type TemporalBufferKey = (ScalarImpl, Row);
```

Comment:
Actually, if we make the PK always prefixed by the temporal filter key, we just need this to be a `Row`. Since we need to serialize the PK for insertion, this may as well be `Vec<u8>` (but computing vnode upon scaling/update will be more challenging).